Add Rohde&Schwarz HMP4040 power supply.

bleykauf opened the pull request:
I added most commands for the Rohde&Schwarz HMP4040 with the exception of the fuse commands (since I could not test those commands). 

For some commands I implemented both `Instrument.control` and defined a callable member function since it seems a little odd to for example select a channel via `hmp.selected_channel = 1`; I think it is more natural to do this via `hmp.select_channel(1)`. These commands are annotated in the source code. Let me know what you think :)

Manual for the device for reference: https://scdn.rohde-schwarz.com/ur/pws/dl_downloads/dl_common_library/dl_manuals/gb_1/h/hmp_serie/HMP_GettingStarted_en_02.pdf

b-leconte commented:
Hello,
 
Just tested your patch.
It is a good thing to have this instrument available in pymeasure.

I had small issue while testing the instrument:
using the function `def activate_channel(self, channel):` is causing an exception.
But if i comment the lines in this function:
`"selected_channel = self.selected_channel" `
and
`"self.selected_channel = selected_channel"` 
Then there is no issue anymore (except that you loose the backup of previously selected channel).

Regards,

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        values=[0, 32.050],
+    )
+
+    def voltage_up(self):
```
better `step_voltage_up`?

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        """Increase voltage by one step."""
+        self.write("VOLT UP")
+
+    def voltage_down(self):
```
better `step_voltage_down`?

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        "CURR:STEP?", "CURR:STEP %s", "Current step in A."
+    )
+
+    def current_up(self):
```
better `step_current_up`?

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        """Increase current by one step."""
+        self.write("CURR UP")
+
+    def current_down(self):
```
better `step_current_down`?

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        map_values=True,
+    )
+
+    output_on = Instrument.control(
```
`output_enabled` would echo some other instruments that use that name already (there is none that has `output_on`)

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        # Restore previously selected channel.
+        self.selected_channel = selected_channel
+
+    # Convenience functions to make turning on/off the output more natural.
+    def turn_output_on(self):
+        """Turn output on."""
+        self.output_on = True
+
+    def turn_output_off(self):
+        """Turn output off."""
+        self.output_on = False
```
I have to disagree that these are "convenience functions". Also, in a pymeasure context, using properties is actually more "natural".
While we have methods in our instruments to do things that are difficult/complex with properties, we do not just mirror functionality that already exists in properties, as this is just confusing for the users. I'd rather you remove those.

bleykauf replied:
Removed the mirrored functionality.

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+
+    def clear_sequence(self, channel):
+        """Clear the sequence of the selected channel."""
+        channel = strict_discrete_set(channel, [1, 2, 3, 4])
```
good way of using our validators!

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        "ARB:REP?",
+        "ARB:REP %s",
+        "Number of repetitions (0...255). If 0 is entered, the sequence is"
+        "repeated infinitely.",
```
*indefinitely

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        "repeated infinitely.",
+        validator=strict_discrete_set,
+        values=range(256),
+        get_process=lambda x: int(x),
```
this can be replaced by `cast=int`.

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        self.write(f"ARB:REST {slot}")
+
+    def save_sequence(self, slot):
+        """Save the sequence to internal memory (slot 1, 2 or 3)."""
```
I guess this means the "current"/active sequence (that has been defined with `sequence`)? Should we point this out in the docstring?

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+    # The following commands are for making it easier to change selected the
+    # selected channel and activate/deactivate them.
+
+    def activate_channel(self, channel):
```
See how this and the following method duplicate all code except one Boolean? This could be united into one method, e.g. `def set_channel_active(self, channel, state):`

TBH, I am a bit unclear about the meanings of the different terms in the manual - "select", "activate" and "turn on output" seem to be three different things. It's clear that "select" designates a channel for some subsequent operations, but the difference between "activating" a (selected) channel and turning its output on is not clear to me (manual p 69)

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+        get_process=lambda x: int(x),
+    )
+
+    # Implemented as a more natural way of selecting a channel.
```
Actually, in our usage a method is not "more natural" than a property. In any case, having both purely for duplicating functionality is confusing, so I think the method should be dropped.

bleykauf replied:
Removed duplicated functionality.

bilderbuchi reviewed (changes requested):
Thank you for your effort! I reviewed the PR and left suggestions/change requests.

bleykauf commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,288 @@
+    # turn sequence data into a string
+    sequence = str(sequence).strip('[](){}')
+    sequence.replace(" ", "")
+    print(sequence)
```
Simply missed to remove this, no need for debug log I think.

bleykauf commented:
I also added some instrument tests, i.e. tests where an instrument is necessary.

For this, I modified `conftest.py` and added a new marker `needs_instrument` and the possibility to specify the resource address via the command line with an command line option `--resource-name`. There is also a command line flag `--instrument-present` to indicate that an instrument is present (this might actually be a little redundant and I am not 100% happy with the way it is right now, will have another look soon).

I think in general pymeasure could benefit from such a functionality since it would allow to clearly separate tests that could be run during CI and tests that need an instrument with the possibility so specify the resource name from the command line so everybody with a specific device can run the tests. The existing tests could be adjusted accordingly.

What do you think?

bleykauf commented:
Opened an issue about the instrument specific tests, #632.

I would also be happy to work on this in a separate PR since it is out of scope for this PR (that just wants to add an instrument).

bilderbuchi commented:
> I would also be happy to work on this in a separate PR since it is out of scope for this PR (that just wants to add an instrument).

Agreed, please do split this out. In general PRs should have as small a scope as useful, as the reviewing effort goes up enormously with diff size.

bleykauf commented:
Simplified the tests and moved them to a separate file that is skipped by default. Ready for re-review :)

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,271 @@
+log.addHandler(logging.NullHandler())
+
+
+def process_sequence(sequence: List[float]) -> str:
```
We [prefer](https://pymeasure.readthedocs.io/en/latest/dev/coding_standards.html#python-style-guides) (probably mainly for historical reasons) to put type hints into the (sphinx-style, like e.g. [here](https://github.com/pymeasure/pymeasure/blob/8086835d8a211268a26451c28a054770b3907d40/pymeasure/instruments/keithley/keithley2400.py#L387)) docstrings, not the signature/code.
Also in `set_channel_state`

bleykauf replied:
Moved the type hints to the docstring.

I think, however, that using the built-in type hints should be allowed. in the future. pymeasure is now Python 3.7+, they are less verbose and easier to read in plain-text format than sphinx-style docstrings and can be checked by your IDE. There is also a plugin for sphinx to make use of type hints: https://pypi.org/project/sphinx-autodoc-typehints/

bilderbuchi replied:
Good to know, thanks!

bilderbuchi reviewed (changes requested):
One minor thing, then we're done here :-)

bilderbuchi commented on pymeasure/instruments/rohdeschwarz/hmp.py:
```diff
@@ -0,0 +1,271 @@
+class HMP4040(Instrument):
+    """Represents a Rohde&Schwarz HMP4040 power supply."""
+
+    def __init__(self, resourceName, name="Rohde&Schwarz HMP4040", **kwargs):
+        super().__init__(
+            resourceName, name, includeSCPI=True, **kwargs
```
```suggestion
    def __init__(self, adapter, **kwargs):
        kwargs.setdefault("name", "Rohde&Schwarz HMP4040")
        super().__init__(
            adapter, name, includeSCPI=True, **kwargs
```
anticipating some changes already in the pipeline :-)
